solr/core/src/java/org/apache/solr/cloud/Overseer.java

```diff
@@ -775,6 +779,42 @@ private void doCompatCheck(BiConsumer<String, Object> consumer) {
     }
   }
 
+  /**
+   * Start {@link ClusterSingleton} plugins when we become the leader.
+   */
+  private void startClusterSingletons() {
+    PluginBag<SolrRequestHandler> handlers = getCoreContainer().getRequestHandlers();
+    if (handlers == null) {
+      return;
+    }
+    handlers.keySet().forEach(handlerName -> {
+      SolrRequestHandler handler = handlers.get(handlerName);
+      if (handler instanceof ClusterSingleton) {
+        try {
+          ((ClusterSingleton) handler).start();
+        } catch (Exception e) {
+          log.warn("Exception starting ClusterSingleton " + handler, e);
+        }
+      }
+    });
+  }
+
+  /**
+   * Stop {@link ClusterSingleton} plugins when we lose leadership.
+   */
+  private void stopClusterSingletons() {
+    PluginBag<SolrRequestHandler> handlers = getCoreContainer().getRequestHandlers();
```

Comment:
Should we stop currently configured `ClusterSingleton` handlers or rather stop all those we've started? Is the configuration immutable?

Reply:
I don't think it matters what was the configuration - the contract for `ClusterSingleton` requires that we stop all of them.